src/os_auth/auth.c

```diff
@@ -202,24 +209,55 @@ w_err_t w_auth_replace_agent(keyentry *key,
+        return OS_INVALID;
+    }
+
+    j_date_add = cJSON_GetObjectItem(j_agent_info->child, "date_add");
+    j_connection_status = cJSON_GetObjectItem(j_agent_info->child, "connection_status");
+    j_disconnected_time = cJSON_GetObjectItem(j_agent_info->child, "disconnected_time");
+
     /* Check if the agent has been disconnected longer than the value specified in the configuration option*/
     if (force_options->disconnected_time_enabled) {
-        time_t disconnected_time = 0;
-        disconnected_time = (long)get_time_since_agent_disconnection(key->id);
-        if (!disconnected_time) {
-            minfo("Agent '%s' can't be replaced since it is not disconnected.", key->id);
+        if (!cJSON_IsString(j_connection_status)) {
+            mdebug1("Failed to get connection status for agent '%s'.", key->id);
+            cJSON_Delete(j_agent_info);
             return OS_INVALID;
-        } else if (disconnected_time > 0 && disconnected_time < force_options->disconnected_time) {
-            minfo("Agent '%s' has not been disconnected long enough to be replaced.", key->id);
+        }
+        if (!cJSON_IsNumber(j_disconnected_time)) {
+            mdebug1("Failed to get disconnected time for agent '%s'.", key->id);
+            cJSON_Delete(j_agent_info);
             return OS_INVALID;
         }
+
+        if (j_connection_status->valuestring && strcmp(j_connection_status->valuestring, AGENT_CS_NEVER_CONNECTED)) {
+            time_t time_since_disconnected = difftime(time(NULL), j_disconnected_time->valueint);
+            if (!strcmp(j_connection_status->valuestring, AGENT_CS_DISCONNECTED) && j_disconnected_time->valueint > 0 && time_since_disconnected < force_options->disconnected_time) {
+                minfo("Agent '%s' has not been disconnected long enough to be replaced.", key->id);
+                cJSON_Delete(j_agent_info);
+                return OS_INVALID;
+            } else if (j_disconnected_time->valueint == 0) {
+                minfo("Agent '%s' can't be replaced since it is not disconnected.", key->id);
+                cJSON_Delete(j_agent_info);
+                return OS_INVALID;
+            }
+        }
     }
 
     /* Check if the agent is old enough to be removed */
-    if(force_options->after_registration_time > 0) {
-        time_t agent_registration_time = get_time_since_agent_registration(atoi(key->id));
-        if(agent_registration_time > 0 && agent_registration_time <= force_options->after_registration_time){
+    if (force_options->after_registration_time > 0) {
+        if (!cJSON_IsNumber(j_date_add)) {
+            mdebug1("Failed to get registration time for agent '%s'.", key->id);
+            cJSON_Delete(j_agent_info);
+            return OS_INVALID;
+        }
+
+        time_t agent_registration_time = difftime(time(NULL), j_date_add->valueint);
+
+        if (agent_registration_time > 0 && agent_registration_time < force_options->after_registration_time) {
```

Comment:
I think it isn´t required to check **agent_registration_time > 0**.
We can rename **agent_registration_time** to **agent_time_since_desconnection**

Reply:
Changed